keps/sig-network/2091-cluster-network-policy/README.md

```diff
@@ -0,0 +1,969 @@
+1. Logging / error reporting for Network Policy
+2. Kubernetes Node policies
+3. New policy selectors (services, service accounts, etc.)
+4. Tools/CLI to debug/explain NetworkPolicies, cluster-scoped policies and their
+   impact on workloads.
+
+## Proposal
+
+In order to achieve the two primary broad use cases for a cluster admin to
+secure K8s clusters, we propose to introduce the following two new resources
+under `networking.k8s.io` API group:
+- ClusterNetworkPolicy
+- DefaultNetworkPolicy
+
+### ClusterNetworkPolicy resource
+
+A ClusterNetworkPolicy resource will help the administrators set strict
+security rules for the cluster, i.e. a developer CANNOT override these rules
+by creating NetworkPolicies that applies to the same workloads as the
+ClusterNetworkPolicy does.
+
+Unlike the NetworkPolicy resource in which each rule represents a allowed
+traffic, ClusterNetworkPolicy will enable administrators to set `Allow` or
+`Deny` as the action of each rule. ClusterNetworkPolicy rules should be read
+as-is, i.e. there will not be any implicit isolation effects for the Pods
+selected by the ClusterNetworkPolicy, as opposed to what NetworkPolicy rules imply.
+
+In terms of precedence, the aggregated `Deny` rules (all ClusterNetworkPolicy
+rules with action `Deny` in the cluster combined) should be evaluated before
+aggregated ClusterNetworkPolicy `Allow` rules, followed by aggregated
+NetworkPolicy rules in all Namespaces. As such, the `Deny` rules have the
+highest precedence, which prevents them to be unexpectedly overwritten.
+
+ClusterNetworkPolicy `Deny` rules are useful for administrators to explicitly
+block traffic from malicious clients, or workloads that poses security risks.
+Those traffic restrictions can only be lifted once the`Deny` rules are deleted
+or modified. On the other hand, the `Allow` rules can be used to call out
+traffic in the cluster that needs to be allowed for certain components to work
+as expected (egress to CoreDNS for example). Those traffic should not be blocked
+when developers apply NetworkPolicy to their Namespaces which isolates the workloads.
+
+### DefaultNetworkPolicy resource
+
+A DefaultNetworkPolicy resource will help the administrators set baseline security
+rules for the cluster, i.e. a developer CAN override these rules by creating
+NetworkPolicies that applies to the same workloads as the DefaultNetworkPolicy does.
+
+DefaultNetworkPolicy works just like NetworkPolicy except that it is cluster-scoped.
+When workloads are selected by a DefaultNetworkPolicy, they are isolated except
+for the ingress/egress rules specified. DefaultNetworkPolicy rules will not have
+actions associated -- each rule will be an 'allow' rule.
+
+Aggregated NetworkPolicy rules will be evaluated before aggregated DefaultNetworkPolicy rules.
+If a Pod is selected by both, a DefaultNetworkPolicy and a NetworkPolicy, then
+the DefaultNetworkPolicy's effect on that Pod becomes obsolete.
```

Comment:
It feels a bit counter-intuitive to me that the NetworkPolicy "default deny" gets implemented before the DefaultNetworkPolicy - doesn't this render DNPs somewhat useless? I think I might not be grokking what the purpose of DNP is.

Reply:
I would have expected:

1. NetworkPolicy rules
2. DefaultNetworkPolicy rules
3. Default deny

Reply:
We gave this a bit of thought before opening the KEP, and I believe the reason we wanted NetworkPolicy default deny to have higher priority over DefaultNetworkPolicy rules is that, Namespace users will not be able to overwrite DefaultNetworkPolicy rules otherwise (or maybe this is what's expected in most cases based on previous comments?), so that statement about DefaultNetworkPolicy rules become obsolete when user apply NetworkPolicy becomes a lie

Reply:
The implicit default-deny-on-selected behavior being different across the different policy types feels like a major area for confusion to me. It seems to me that the most intuitive behavior would be that whenever _any_ type of policy selects a pod, it's default behavior goes from allow to deny.

Then, it's simply a matter of rule ordering.

1. cluster-exceptions
2. cluster-deny
3. cluster-allow
4. namespaced NP rules.
5. default-deny.

This does preclude writing more "complex" DefaultNetworkPolicies, but I have yet to see a real use-case for a DNP. The closest thing is "default to deny anything except within my namespace", but I think that can be implemented without DNP.

e.g., 

1. A CNP that selects everything, allows nothing (thus creating a default-deny for all pods).
2. a NP in each namespace that allows within the NS. This could, of course, be edited or removed by the users within that NS, but that's no different than the "overridding" mechanics present in the current DNP proposal.

Reply:
>  It seems to me that the most intuitive behavior would be that whenever _any_ type of policy selects a pod, it's default behavior goes from allow to deny.

We decided that the CNP actions should be 'as-is' because the semantics is less implicit this way - it has a couple of major benefits:
- Each allowed/dropped traffic matching a CNP will be attributed to a specific rule, rather than implied by some isolation effects
- It makes the meaning of the CNP actions more 'pure'. If these actions come with default isolation, then a standalone CNP `Deny` rule won't make any sense: isolate ingress for workload xxx and deny ingress from yyy specifically?

> A NP in each namespace that allows within the NS. This could, of course, be edited or removed by the users within that NS

In this KEP use think the Namespace (as tenant boundary) isolation should be the persona of a cluster admin rather than a namespace user. Of course the same traffic enforcements can be achieved by NetworkPolicies, but as you pointed out, the developers in each Namespace would have to maintain essentially the same policy, which should come as a cluster default for them.

Reply:
I tend to consider consistency between similar APIs to be one-of, if not _the_ top consideration when building a new API. If this were _replacing_ NetworkPolicy, we'd have a bit more freedom IMO. But this is an API that is meant to work alongside NetworkPolicy, and so having subtle behavioral/syntactic/etc. differences is an easy way to make the APIs harder for users to get their heads around.

Reply:
I do agree that consistency is a very important factor. As a matter of fact, for the CRDs that we plan to introduce to the new repo (which would host the ClusterNetworkPolicy resources and other improvements to NetworkPolicy), consistency is our main concern. This is why @abhiraut  @astoycos @rikatz and I believe CNP might be a good start for us to get rid of the implicit isolation behavior, even though it will need to co-exist with the current NetworkPolicy for a while. If new semantics makes sense, then it can become the guiding principle for the new CRDs, which eventually _can_ replace the current NetworkPolicy

Reply:
@caseydavenport Network Policy Was originally designed to express the Developer's Intent, while the main persona in this proposal is the administrator so point being.....shouldn't 90% of design choices be done with the main user in mind? If so then why are we following precedents set by Network Policy, it's like trying to put a square peg in a round hole. 

We know Network Policy is here to stay, and is a great tool for the developer to apply security policies, however I would argue it is not a great tool for an administrator for a number of reasons, one of them being the implicit isolation behavior. 

I guess my main point is that we're starting the new `network-policy-api` repo to house the CNP and a possible new object around Network Policy (I hate to call it V2 :) in a manner that won't break any existing API.  Therefore, If the community decides that the new personas/ user stories for such objects are not in line with Existing NP then is there really anything holding us back from breaking away from the conventions used previously?

Reply:
> shouldn't 90% of design choices be done with the main user in mind? If so then why are we following precedents set by Network Policy, it's like trying to put a square peg in a round hole.

While we talk about two distinct personas in the proposal, in the real-world these roles may be implemented by the same person or people within a business. Or they might even be split among more than two different people/groups.

> I hate to call it V2 ... Therefore, If the community decides that the new personas/ user stories for such objects are not in line with Existing NP then is there really anything holding us back from breaking away from the conventions used previously?

Yep, this is exactly what I am trying to explain. We're not building a v2. We're not replacing the existing API. We're augmenting it, and the new API(s) should behave consistently and intuitively with the APIs that already exist in v1. 

Even if one makes the argument that the existing default-deny mechanics are confusing, adding another API that behaves differently won't fix that, and will instead just make the set of APIs even more confusing.

If we ever do build a v2 NetworkPolicy, then that should be its own effort separate from this ClusterNetworkPolicy proposal, and that would be the time we should consider making changes across the board to NP / CNP / DNP / whatever NP APIs exist. But we would still aim for the set of v2 APIs to be consistent with each other. 

I'm not the only person who was caught off-guard by this unexpected difference: https://github.com/kubernetes/enhancements/pull/2522#discussion_r617282128

The responses I received above talk mostly in vague terms like "API purity", "less implicit", etc. Is there a cut-and-dry use-case that necessitates this? Because that's a lot more convincing of an argument. In other words, we need to justify making this behavior change with a use-case rather than the other way around.

Reply:
> The implicit default-deny-on-selected behavior being different across the different policy types feels like a major area for confusion to me

I agree with this 100%, though I disagree on the fix; I think we should deprecate NetworkPolicy and replace it with a new ApplicationNetworkPolicy that is mostly similar, but has only explicit deny semantics.

Implicit default-deny-on-selected is the single most terrible decision sig-network has ever made (which is understandable given that we made it under extreme time pressure after discovering a day or two before feature freeze that we weren't going to be allowed to turn our Namespace annotation into a Namespace field).

> It seems to me that the most intuitive behavior would be that whenever _any_ type of policy selects a pod, it's default behavior goes from allow to deny.
> 
> Then, it's simply a matter of rule ordering.
> 
>     1. cluster-exceptions
>     2. cluster-deny
>     3. cluster-allow
>     4. namespaced NP rules.
>     5. default-deny.

That means there's no way to create a policy saying "allow ingress from prometheus to all namespaces" without causing the entire cluster to become isolated. There is no way to just say "I want to allow prometheus to reach all pods". You have to say "I want to allow prometheus to reach all pods, and I also want to block all other traffic".

Reply:
> I agree with this 100%, though I disagree on the fix; I think we should deprecate NetworkPolicy and replace it with a new ApplicationNetworkPolicy that is mostly similar, but has only explicit deny semantics.

Yep, I think this would also be a fine approach, ans is basically aligned with what I was saying above - either we need to build something that plays nice with NetworkPolicy, or we need to build something entirely new that replaces NetworkPolicy, but in either case we need consistency. 

> That means there's no way to create a policy saying "allow ingress from prometheus to all namespaces" without causing the entire cluster to become isolated.

Yeah, I had this realization the other day as well. I think this is probably why the proposal is the way that it is currently. I suspect there are other ways to resolve this, but they might be even less desirable.

Reply:
> something entirely new that replaces NetworkPolicy

FWIW, we could define "ApplicationNetworkPolicy" in a way that would still be compatible with NetworkPolicy. Just give it explicit Allow and Deny rules, where the Allows have priority over the Denies, and then say that a NetworkPolicy that says

    ingress:
      - from:
          - X
          - Y
          - Z

is equivalent to an ApplicationNetworkPolicy that says

    ingress:
      - action: Allow
          - from:
              - X
              - Y
              - Z
      - action: Deny
          - from:
              - ipBlock:
                  cidr: 0.0.0.0/0

Then it's not difficult to support both new ApplicationNetworkPolicies and legacy NetworkPolicies at the same time, and it's always clear how legacy NetworkPolicies interact with CNPs/DNPs, etc (although if we were going to do this then this would presumably involve changes to how CNP and DNP were defined too.)

Reply:
>While we talk about two distinct personas in the proposal, in the real-world these roles may be implemented by the same person or people within a business. Or they might even be split among more than two different people/groups.

I agree, but at the same time we need to design for one of them... Trying to "optimize" a design for two different personas is what I think has been causing some of the confusion in this KEP. 

Overall the question of what API this fits into definitely needs to be solidified.  

@danwinship I like that idea, there was actually some discussion about having the same sort of thing in CNP (i.e two actions for a CNP/DNP  with allows overriding denys) and I think it's something that _could_ be fit into the existing API. 

@Dyanngg This ^ is sort of what I was referring to in the meeting today. The decision about how actions will work with CNP (i.e the 3 actions, only allow/deny, rules with priority, etc. ) Will ultimately be a major factor in the question of are we trying to "augment" the existing API or begin forming a new set of "v2" APIs. Specifically something like what Dan is proposing could "fit" into the existing API while something like actions with configurable priorities may not. 

Hopefully after further analysis of the main personas and core user stories it will be a bit easier to draw some conclusions here.